tests/test_scheduler.py

```diff
@@ -0,0 +1,136 @@
+"""Unit tests for scheduler module."""
+
+import unittest
+from unittest import mock
+
+from PyQt5.QtWidgets import QApplication
+from PyQt5.QtCore import Qt, QObject,  QMimeData
+
+from iquip.apps import scheduler
+from iquip.protocols import ExperimentInfo
+
+class ExperimentModelTest(unittest.TestCase):
+    """Unit tests for ExperimentModel class."""
+
+    def setUp(self):
+        self.qapp = QApplication([])
+
+    def tearDown(self):
+        del self.qapp
+
+    def test_model_index(self):
+        data1 = []
+        for i in range(10):
+            data1.append(ExperimentInfo(str(i), {"rid": i, "priority": i}))
+        data2 = []
+        for i in range(100):
+            data1.append(ExperimentInfo(str(i), {"rid": i, "priority": 0}))
+        for data in (data1, data2):
+            mdl = scheduler.ExperimentModel()
+            mdl.experimentQueue.extend(data)
+            self.assertEqual(mdl.rowCount(), len(data))
+            for i, exp in enumerate(data):
+                self.assertEqual(mdl.data(mdl.index(i)), exp)
+
+    def test_drag_and_drop(self):
+        mdl = scheduler.ExperimentModel()
+        data = [ExperimentInfo("1", {"rid": 1, "priority": 2}),
+                ExperimentInfo("2", {"rid": 2, "priority": 1}),
+                ExperimentInfo("3", {"rid": 3, "priority": 1})
+               ]
+        mdl.experimentQueue.extend(data)
+        mime0 = QMimeData()
+        mime0.setText("0")
+        mime1 = QMimeData()
+        mime1.setText("1")
+        mime2 = QMimeData()
+        mime2.setText("2")
+        mdl.dropMimeData(mime0, Qt.MoveAction, 0, 0, mdl.index(0)) # exp1 above exp1
+        self.assertEqual(mdl.experimentQueue, data)
+        mdl.dropMimeData(mime0, Qt.MoveAction, 2, 0, mdl.index(0)) # exp1 above exp3
+        self.assertEqual(mdl.experimentQueue, data)
+        mdl.dropMimeData(mime1, Qt.MoveAction, 3, 0, mdl.index(0)) # exp2 below exp3
+        self.assertEqual(mdl.experimentQueue, [data[0], data[2], data[1]])
+        mdl.dropMimeData(mime2, Qt.MoveAction, 1, 0, mdl.index(0)) # exp2 above exp3
+        self.assertEqual(mdl.experimentQueue, data)
+
+
+class SchedulerAppTest(unittest.TestCase):
+    """Unit tests for SchedulerApp class."""
+
+    def setUp(self):
+        self.qapp = QApplication([])
+
+    def tearDown(self):
+        del self.qapp
+
+    def test_add_experiment(self):
+        app = scheduler.SchedulerApp(name="name")
+        with mock.patch.object(app.schedulerFrame.model, "experimentQueue") as mockedQueue:
+            data = [ExperimentInfo(str(i), {"rid": i, "priority": 10 - i}) for i in range(10)]
+            for info in data:
+                app.addExperiment(info)
+                mockedQueue.append.assert_called_with(info)
+
+    def test_run_experiment(self):
```

Comment:
Please test the case when the added experiment was in the queue, as another test method, e.g., `test_run_experiment_delete_from_queue()`.

Reply:
I found out that there is no situation that the added experiment was in the queue, as it receives the entire queue every time. There is no signal emitted when the experiment is run. Do I need to make this test even if it will be deleted afterward?

Reply:
Do you mean that you are going to remove the code (in `scheduler.py`) which checks if the experiment is already in the queue?

Reply:
Yes.